File: leetcode/array/1198_Find_Smallest_Common_Element_in_All_Rows/codes/bs.c
#include <stdbool.h>

bool binarySearch(int* arr, int size, int target) {
    int l = 0, r = size - 1;
    while (l <= r) {
        int mid = (l + r) / 2;
        if (arr[mid] == target) return true;
        else if (arr[mid] < target) l = mid + 1;
        else r = mid - 1;
    }
    return false;
}

int smallestCommonElement(int** mat, int matSize, int* matColSize) {
    int firstRowSize = matColSize[0];

    for (int k = 0; k < firstRowSize; ++k) {
        int num = mat[0][k];
        bool foundInAll = true;

        for (int i = 1; i < matSize; ++i) {
            if (!binarySearch(mat[i], matColSize[i], num)) {
                foundInAll = false;
                break;
            }
        }

        if (foundInAll) return num;
    }

    return -1;
}

File: leetcode/array/1198_Find_Smallest_Common_Element_in_All_Rows/codes/bs.py
from typing import List

class Solution:
    def smallestCommonElement(self, mat: List[List[int]]) -> int:
        def binary_search(arr, target):
            l, r = 0, len(arr) - 1
            while l <= r:
                mid = (l + r) // 2
                if arr[mid] == target:
                    return True
                elif arr[mid] < target:
                    l = mid + 1
                else:
                    r = mid - 1
            return False

        for num in mat[0]:
            if all(binary_search(row, num) for row in mat[1:]):
                return num
        return -1
